src/mongocxx/options/find.cpp

```diff
@@ -211,6 +214,179 @@ const stdx::optional<class read_preference>& find::read_preference() const {
+                    }
+                    break;
+                default:
+                    throw exception{error_code::k_invalid_parameter,
+                                    "document or string type required for $hint"};
+            }
+        } else if (ele.key() == stdx::string_view("$max")) {
+            if (ele.type() != bsoncxx::type::k_document) {
+                throw exception{error_code::k_invalid_parameter, "document type required for $max"};
+            }
+            if (!_max) {
+                _max = bsoncxx::document::value{ele.get_document().value};
+            }
+        } else if (ele.key() == stdx::string_view("$maxScan")) {
+            std::int32_t max_scan_int32;
+            std::int64_t max_scan_int64;
+            double max_scan_double;
+            switch (ele.type()) {
+                case bsoncxx::type::k_int32:
+                    max_scan_int32 = ele.get_int32().value;
+                    if (max_scan_int32 < 0) {
+                        throw exception{error_code::k_invalid_parameter,
+                                        "non-negative value required for $maxScan"};
+                    }
+                    if (!_max_scan) {
+                        _max_scan = max_scan_int32;
+                    }
+                    break;
+                case bsoncxx::type::k_int64:
+                    max_scan_int64 = ele.get_int64().value;
+                    if (max_scan_int64 < 0 ||
+                        max_scan_int64 > std::numeric_limits<std::int32_t>::max()) {
+                        throw exception{error_code::k_invalid_parameter,
+                                        "value for $maxScan required to be between 0 and 2^31-1"};
+                    }
+                    if (!_max_scan) {
+                        _max_scan = static_cast<std::int32_t>(max_scan_int64);
+                    }
+                    break;
+                case bsoncxx::type::k_double:
+                    max_scan_double = ele.get_double().value;
+                    if (max_scan_double < 0 ||
+                        max_scan_double > std::numeric_limits<std::int32_t>::max()) {
+                        throw exception{error_code::k_invalid_parameter,
+                                        "value for $maxScan required to be between 0 and 2^31-1"};
+                    }
+                    if (!_max_scan) {
+                        _max_scan = static_cast<std::int32_t>(max_scan_double);
+                    }
+                    break;
+                default:
+                    throw exception{error_code::k_invalid_parameter,
+                                    "numberic type required for $maxScan"};
+            }
+        } else if (ele.key() == stdx::string_view("$maxTimeMS")) {
+            std::int32_t max_time_ms_int32;
+            std::int64_t max_time_ms_int64;
+            double max_time_ms_double;
+            switch (ele.type()) {
```

Comment:
Same issue here and having such long code duplicated makes understanding what's going on harder.